satellite/gracefulexit/endpoint_test.go

```diff
@@ -1207,6 +1207,138 @@ func TestFailureNotFoundPieceHashUnverified(t *testing.T) {
+	var maxOrderLimitSendCount = 3
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount:   1,
+		StorageNodeCount: 5,
+		UplinkCount:      1,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(logger *zap.Logger, index int, config *satellite.Config) {
+				// We don't care whether a node gracefully exits or not in this test,
+				// so we set the max failures percentage extra high.
+				config.GracefulExit.OverallMaxFailuresPercentage = 101
+				config.GracefulExit.MaxOrderLimitSendCount = maxOrderLimitSendCount
+			},
+		},
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		uplinkPeer := planet.Uplinks[0]
+		satellite := planet.Satellites[0]
+
+		satellite.GracefulExit.Chore.Loop.Pause()
+
+		nodeFullIDs := make(map[storj.NodeID]*identity.FullIdentity)
+		for _, node := range planet.StorageNodes {
+			nodeFullIDs[node.ID()] = node.Identity
+		}
+
+		rs := &uplink.RSConfig{
+			MinThreshold:     2,
+			RepairThreshold:  3,
+			SuccessThreshold: 4,
+			MaxThreshold:     4,
+		}
+
+		err := uplinkPeer.UploadWithConfig(ctx, satellite, rs, "testbucket", "test/path", testrand.Bytes(5*memory.KiB))
+		require.NoError(t, err)
+
+		// check that there are no exiting nodes.
+		exitingNodes, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
+		require.NoError(t, err)
+		require.Len(t, exitingNodes, 0)
+
+		exitingNode, err := findNodeToExit(ctx, planet, 1)
+		require.NoError(t, err)
+
+		// connect to satellite so we initiate the exit.
+		conn, err := exitingNode.Dialer.DialAddressID(ctx, satellite.Addr(), satellite.Identity.ID)
+		require.NoError(t, err)
+		defer ctx.Check(conn.Close)
+
+		client := conn.SatelliteGracefulExitClient()
+
+		c, err := client.Process(ctx)
+		require.NoError(t, err)
+
+		response, err := c.Recv()
+		require.NoError(t, err)
+
+		// should get a NotReady since the metainfo loop would not be finished at this point.
+		switch response.GetMessage().(type) {
+		case *pb.SatelliteMessage_NotReady:
+			// now check that the exiting node is initiated.
+			exitingNodes, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
+			require.NoError(t, err)
+			require.Len(t, exitingNodes, 1)
+
+			require.Equal(t, exitingNode.ID(), exitingNodes[0].NodeID)
+		default:
+			t.FailNow()
+		}
+		// close the old client
+		require.NoError(t, c.CloseSend())
+
+		// trigger the metainfo loop chore so we can get some pieces to transfer
+		satellite.GracefulExit.Chore.Loop.TriggerWait()
+
+		// make sure all the pieces are in the transfer queue
+		_, err = satellite.DB.GracefulExit().GetIncomplete(ctx, exitingNode.ID(), 1, 0)
+		require.NoError(t, err)
+
+		var messageCount int
+
+		// We need to label this outer loop so that we're able to exit it from the inner loop.
+		// The outer loop is for sending the request from node to satellite multiple times.
+		// The inner loop is for reading the response.
+	MessageLoop:
+		for messageCount <= maxOrderLimitSendCount {
```

Comment:
nit - I think we should exit the loop either with the `messageCount <= maxOrderLimitSendCount` (preferable to me), or change this to a `for {` with no condition and use the `break MessageLoop` on line 1310. I don't think we should have both, because then it is not clear what is causing the exit.
But it's not a big deal so don't worry about it too much.

Reply:
hmm I think the reason for both is because we weren't sure how to break out of both the inner and outer loop without adding the label `MessageLoop` and using `break MessageLoop`. do you know how we could do this the preferred way without using `break MessageLoop`?

Reply:
what I am saying is there are two potential ways to exit here. One is if `messageCount > maxOrderLimitSendCount` (which now that I am thinking about it should never happen). And the other is `break MessageLoop`. The test would be clearer if we only had one of these potential exit routes, so we should pick one and use that, and get rid of the other. So if you want to keep using `MessageLoop`, feel free, but in that case I think you should change the for loop to not have any condition (`for {`)

Reply:
ah okay I understand, thanks